apps/api/src/http/v3/rest/auth/login.rs

```diff
@@ -134,151 +128,81 @@ pub async fn handle_callback(global: &Arc<Global>, query: LoginRequest, cookies:
-								}
-							}
-						},
-						update::update! {
-							#[query(set)]
-							User {
-								#[query(flatten, index = "$")]
-								connections: UserConnection {
-									platform_username: user_data.username,
-									platform_display_name: user_data.display_name,
-									platform_avatar_url: user_data.avatar,
-									updated_at: chrono::Utc::now(),
-								},
-								updated_at: chrono::Utc::now(),
-							}
-						},
-						None,
-					)
-					.await?
-					.matched_count == 0
-				{
-					tracing::error!("failed to update user, no matched count");
-					return Err(TransactionError::custom(ApiError::INTERNAL_SERVER_ERROR));
+		// upsert the connection
+		tx.update_one(
+			filter::filter! {
+				User {
+					#[query(rename = "_id")]
+					id: user_id,
+					#[query(elem_match)]
+					connections: UserConnection {
+						platform,
+						platform_id: &user_data.id,
+					}
 				}
-			}
-		} else {
-			let new_connection = UserConnection {
-				platform,
-				platform_id: user_data.id,
-				platform_username: user_data.username,
-				platform_display_name: user_data.display_name,
-				platform_avatar_url: user_data.avatar,
-				allow_login: true,
-				updated_at: chrono::Utc::now(),
-				linked_at: chrono::Utc::now(),
-			};
-
-			let res = tx
-				.update_one(
-					filter::filter! {
-						User {
-							#[query(rename = "_id")]
-							id: full_user.user.id,
-						}
-					},
-					update::update! {
-						#[query(push)]
-						User {
-							#[query(serde)]
-							connections: &new_connection,
-						},
-						#[query(set)]
-						User {
-							updated_at: chrono::Utc::now(),
-						}
-					},
-					None,
-				)
-				.await?;
-
-			if res.modified_count > 0 {
-				tx.register_event(InternalEvent {
-					actor: Some(full_user.clone()),
-					data: InternalEventData::User {
-						after: full_user.user.clone(),
-						data: InternalEventUserData::AddConnection {
-							connection: new_connection,
-						},
+			},
+			update::update! {
+				#[query(set)]
+				User {
+					#[query(flatten, index = "$")]
+					connections: UserConnection {
+						platform_username: user_data.username,
+						platform_display_name: user_data.display_name,
+						platform_avatar_url: user_data.avatar,
+						updated_at: chrono::Utc::now(),
 					},
-					timestamp: chrono::Utc::now(),
-				})?;
-			} else {
-				tracing::error!("failed to insert user connection, no modified count");
-				return Err(TransactionError::custom(ApiError::INTERNAL_SERVER_ERROR));
-			}
+					updated_at: chrono::Utc::now(),
+				}
+			},
+			UpdateOptions::builder().upsert(true).build(),
+		)
+		.await?;
+
+		Ok(user_id)
+	})
+	.await
+	.map_err(|e| match e {
+		TransactionError::Custom(e) => e,
+		e => {
+			tracing::error!(error = %e, "transaction failed");
+			ApiError::INTERNAL_SERVER_ERROR
 		}
+	})?;
+
+	let full_user = global
+		.user_loader
+		.load(global, user_id)
+		.await
+		.ok()
+		.flatten()
+		.ok_or(ApiError::INTERNAL_SERVER_ERROR)?;
```

Comment:
Why do we do this? Dont we have the user by this point already?

Reply:
Not the `FullUser`

Reply:
Rather use the `.load_user` function and pass the entire user object created via the transaction